include/seqan3/alignment/matrix/detail/coordinate_matrix.hpp

```diff
@@ -0,0 +1,222 @@
+
+public:
+    /*!\name Associated types
+     * \{
+     */
+    //!\brief The value type.
+    using value_type = decltype(std::views::iota(static_cast<size_t>(0), static_cast<size_t>(1))
+                              | std::views::transform(convert_to_matrix_coordinate{m_column_index}));
+    //!\brief The reference type.
+    using reference = value_type;
+    //!\brief The pointer type.
+    using pointer = void;
+    //!\brief The difference type.
+    using difference_type = std::ptrdiff_t;
+    //!\brief The iterator category.
+    using iterator_category = std::forward_iterator_tag;
+    //!\}
+
+    /*!\name Constructor, assignment and destructor
+     * \{
+     */
+    coordinate_matrix_iterator() noexcept = default; //!< Defaulted.
+    coordinate_matrix_iterator(coordinate_matrix_iterator const &) noexcept = default; //!< Defaulted.
+    coordinate_matrix_iterator(coordinate_matrix_iterator &&) noexcept = default; //!< Defaulted.
+    coordinate_matrix_iterator & operator=(coordinate_matrix_iterator const &) noexcept = default; //!< Defaulted.
+    coordinate_matrix_iterator & operator=(coordinate_matrix_iterator &&) noexcept = default; //!< Defaulted.
+    ~coordinate_matrix_iterator() = default; //!< Defaulted.
+
+    /*!\brief Constructs and initialises the iterator with the current column index and the row index marking the end
+     *        of the rows (size of one column).
+     *
+     * \param[in] column_index \copybrief seqan3::detail::coordinate_matrix_iterator::m_column_index
+     * \param[in] end_row_index \copybrief seqan3::detail::coordinate_matrix_iterator::m_end_row_index
+     */
+    explicit coordinate_matrix_iterator(column_index_type<size_t> column_index,
+                                        row_index_type<size_t> end_row_index) noexcept :
+        m_column_index{column_index.get()},
+        m_end_row_index{end_row_index.get()}
+    {}
+    //!\}
+
+    /*!\name Access operators
+     * \{
+     */
+
+    //!\brief Access the pointed-to matrix coordinate column.
+    auto operator*() const
+    {
+        return std::views::iota(static_cast<size_t>(0), m_end_row_index)
+             | std::views::transform(convert_to_matrix_coordinate{m_column_index});
+    }
+    //!\}
+
+    /*!\name Arithmetic operators
+     * \{
+     */
+
+    //!\brief Increments the iterator to the next column.
+    coordinate_matrix_iterator & operator++()
+    {
+        ++m_column_index;
+        return *this;
+    }
+
+    //!\brief Increments the iterator to the next column and returns the iterator pointing to the previous column.
+    coordinate_matrix_iterator operator++(int)
+    {
+        coordinate_matrix_iterator tmp{*this};
+        ++(*this);
+        return tmp;
+    }
+    //!\}
+
+    /*!\name Comparison operators
+     * \{
+     */
+
+    //!\brief Tests whether `lhs == rhs`.
+    friend bool operator==(coordinate_matrix_iterator const & lhs, coordinate_matrix_iterator const & rhs)
```

Comment:
Just following the cpp core guidelines here: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c86-make--symmetric-with-respect-to-operand-types-and-noexcept

In fact, the != operator can also be non-friend non-member function.

Reply:
Okay this conversion sequence thing might be an argument, but it might also be a counter argument, because we nearly never use any type hierarchy that expects to be convertible. I'm a bit torn with this, because I had to fix more bad friend operators when I ever could actually use. This might be my personal bias with them.